src/tools/perf/lib/libperf.c

```diff
@@ -1028,6 +1034,71 @@ ucp_perf_test_rkey_unpack(ucx_perf_thread_context_t *thread,
     return UCS_OK;
 }
 
+static ucs_status_t ucp_perf_test_create_self_rkey(ucx_perf_context_t *perf,
+                                                   ucp_mem_h memh,
+                                                   ucp_rkey_h *rkey)
+{
+    void *rkey_buffer = NULL;
+    size_t rkey_size  = 0;
+    ucs_status_t status;
+
+    status = ucp_rkey_pack(perf->ucp.context, memh, //perf->ucp.recv_memh,
+                           &rkey_buffer, &rkey_size);
+    if (status != UCS_OK) {
+        if (perf->params.flags & UCX_PERF_TEST_FLAG_VERBOSE) {
+            ucs_fatal("ucp_rkey_pack() failed: %s", ucs_status_string(status));
+        }
+        goto out;
+    }
+
+    status = ucp_ep_rkey_unpack(perf->ucp.self_ep, rkey_buffer, rkey);
+    if (status != UCS_OK) {
+        if (perf->params.flags & UCX_PERF_TEST_FLAG_VERBOSE) {
+            ucs_fatal("ucp_rkey_unpack() failed: %s",
+                      ucs_status_string(status));
+        }
+    }
+
+out:
+    ucs_free(rkey_buffer);
+    return status;
+}
+
+static ucs_status_t ucp_perf_test_setup_self_endpoint(ucx_perf_context_t *perf)
+{
+    ucp_worker_h worker = perf->ucp.worker;
+    size_t address_length;
+    ucp_address_t *address;
+    ucs_status_t status;
+    ucp_ep_params_t ep_params;
+
+    status = ucp_worker_get_address(worker, &address, &address_length);
+    if (status != UCS_OK) {
+        ucs_error("ucp_worker_get_address() failed: %s",
+                  ucs_status_string(status));
+        return status;
+    }
+    ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
+    ep_params.address    = address;
+
+    status = ucp_ep_create(worker, &ep_params, &perf->ucp.self_ep);
+    if (status != UCS_OK) {
+        if (perf->params.flags & UCX_PERF_TEST_FLAG_VERBOSE) {
+            ucs_error("ucp_ep_create() failed: %s", ucs_status_string(status));
+        }
+        return status;
+    }
+
+    status = ucp_perf_test_create_self_rkey(perf, perf->ucp.send_memh,
+                                            &perf->ucp.self_send_rkey);
+    if (status != UCS_OK) {
+        return status;
```

Comment:
need cleanup flow if some stage fails (with goto)

Reply:
But the calling function will do a cleanup for any `status != UCS_OK`:
![image](https://github.com/openucx/ucx/assets/19361875/cc356bfd-f945-4d35-bc76-733682542591)

Reply:
nevertheless, each function should cleanup after itself if it fails